src/tracking.F90

```diff
@@ -85,20 +119,121 @@ subroutine transport(p)
+      ! stored tracking distance, let the particle cross domain once more.
+      ! Note the particle coordinates are recovered back.
+      if (dd_run .and. p % stored_distance > ZERO) then
+        xyz_temp = p % coord(1) % xyz
+        p % coord(1) % xyz = xyz_temp + p % fly_dd_distance * p % coord(1) % uvw
+        call distance_to_mesh_surface(p, domain_decomp % mesh, &
+             distance, d_dd_mesh, meshbin=domain_decomp % meshbin)
+        p % coord(1) % xyz = xyz_temp
+        p % fly_dd_distance = p % fly_dd_distance + d_dd_mesh
+        if (p % stored_distance - p % fly_dd_distance > FP_COINCIDENT) then
+          ! Prepare particle for communication and stop tracking it
+          call cross_domain_boundary(p, domain_decomp, p % stored_distance, &
+               p % fly_dd_distance)
+
+          ! Keep on tracking secondary particles
+          if (p % n_secondary > 0) then
+            call p % initialize_from_source(p % secondary_bank(p % n_secondary))
+            p % n_secondary = p % n_secondary - 1
+            n_event = 0
+
+            ! Set the random number seed and clear crossing data for dd
+            if (dd_run) then
+              prn_seed = p % prn_seed
+              micro_xs % last_E = ZERO
+              micro_xs % last_index_sab = NONE
+              p % sec_particle = .true.
+              n_stage_secondary = n_stage_secondary + 1
+            end if
+
+            ! Enter new particle in particle track file
+            if (p % write_track) call add_particle_track()
+
+            cycle
+          else
+            exit
+          end if
+
+        end if
+      end if
+
       ! Find the distance to the nearest boundary
       call distance_to_boundary(p, d_boundary, surface_crossed, &
            lattice_translation, next_level)
 
       ! Sample a distance to collision
-      if (material_xs % total == ZERO) then
-        d_collision = INFINITY
+      if (dd_run .and. p % stored_distance > ZERO) then
+        if (abs(p % stored_distance - d_boundary) < 100*FP_COINCIDENT) then
+          d_boundary  = p % stored_distance
+          d_collision = INFINITY
+        else
+          d_collision = p % stored_distance
+        end if
+          p % stored_distance = ZERO
+          p % fly_dd_distance = ZERO
       else
-        d_collision = -log(prn()) / material_xs % total
+        if (material_xs % total == ZERO) then
+          d_collision = INFINITY
+        else
+          d_collision = -log(prn()) / material_xs % total
+        end if
       end if
 
       ! Select smaller of the two distances
       distance = min(d_boundary, d_collision)
 
+      ! Check domain mesh boundary
+      if (dd_run) then
+        call distance_to_mesh_surface(p, domain_decomp % mesh, &
+             distance, d_dd_mesh, meshbin=domain_decomp % meshbin)
+      end if
+
+      ! Check for domain boundary crossing
+      if (dd_run .and. d_dd_mesh <= distance) then
```

Comment:
Currently `transport` looks something like this:
```
Check for a domain crossing (exit if so)
Get distance to boundary and collision
Check for a domain crossing (exit if so)
If surface crossing:
  Handle surface crossing
Else (it is a collision):
  Handle collision
Repeat
```

I would prefer it to look more like this
```
Get distance to boundary, collision, and domain crossing
If surface crossing:
  Handle surface crossing
Else if collision:
  Handle collision
Else (it is a domain crossing):
  Handle domain crossing
  (kill the particle rather than exit the loop)
Repeat
```